autoprecompiles/src/optimizer.rs

```diff
@@ -112,6 +118,119 @@ fn solver_based_optimization<T: FieldElement>(
     result.simplified_constraint_system
 }
 
+fn remove_superfluous_constraints<T: FieldElement>(
+    constraint_system: ConstraintSystem<T, Variable>,
+    bus_interaction_handler: impl BusInteractionHandler<T> + Clone + 'static,
+) -> ConstraintSystem<T, Variable> {
+    let (removed_algebraic_constraint_indices, removed_bus_interaction_indices): (
+        HashSet<_>,
+        HashSet<_>,
+    ) = variable_occurrences(&constraint_system)
+        .into_iter()
+        .filter_map(|(v, constraints)| constraints.into_iter().exactly_one().ok().map(|c| (v, c)))
+        .filter_map(|(free_variable, constraint_system_item)| {
+            can_always_solve(
+                &constraint_system,
+                &constraint_system_item,
+                &free_variable,
+                bus_interaction_handler.clone(),
+            )
+            .then_some(constraint_system_item)
+        })
+        .partition_map(|constraint_item| match constraint_item {
+            ConstraintSystemItem::AlgebraicConstraint(i) => Either::Left(i),
+            ConstraintSystemItem::BusInteraction(i) => Either::Right(i),
+        });
+    ConstraintSystem {
+        algebraic_constraints: constraint_system
+            .algebraic_constraints
+            .into_iter()
+            .enumerate()
+            .filter(|(i, _)| !removed_algebraic_constraint_indices.contains(i))
+            .map(|(_, constraint)| constraint)
+            .collect(),
+        bus_interactions: constraint_system
+            .bus_interactions
+            .into_iter()
+            .enumerate()
+            .filter(|(i, _)| !removed_bus_interaction_indices.contains(i))
+            .map(|(_, constraint)| constraint)
+            .collect(),
+    }
+}
+
+fn can_always_solve<T: FieldElement>(
+    constraint_system: &ConstraintSystem<T, Variable>,
+    constraint_system_item: &ConstraintSystemItem,
+    free_variable: &Variable,
+    bus_interaction_handler: impl BusInteractionHandler<T> + Clone + 'static,
+) -> bool {
+    match constraint_system_item {
+        ConstraintSystemItem::AlgebraicConstraint(i) => {
+            // If an algebraic constraint is linear in an otherwise unconstrained variable,
```

Comment:
Can this be seen as dead-code elimination? Basically this variable is not connected to any inputs or outputs?

Reply:
Yeah, true, I think another possible algorithm would be:
 - Build a graph where the nodes are columns and they are connected iff they appear together in a constraint (algebraic or bus interaction)
 - Find connected components
 - Keep only those that are connected to *stateful* bus interactions (execution bridge or memory)

Actually, I guess this would be better, because also cycles would be removed. Maybe we should do this instead. One downside would be that we might remove constraints that can *never* be satisfied, which does change the statement being proven. But maybe it is fair to assume that the constraint system does not contain unsatisfiable constraints?